### ISSUE TITLE
`gsutil cp gs://xxx -` fails with empty error or "ServiceException: 503 None"

### ISSUE THREAD

**Issue by seamusabshere** [faradayio/dbcrossbar#32]
dbcrossbar 0.1.x

700gb data loads are now failing regularly (this error has been seen 4 times in the last 24 hours, each with a different CSV slice)

```
ubuntu@ip-10-0-1-165:/tank$ dbcrossbar cp   --if-exists=append   --schema=postgres-sql:people.sql   gs://xyz/people_v1/   "$DATABASE_URL#people_new"
Failure: ''.
Error: gsutil cp gs://xyz/people_v1/000000000992.csv - failed with exit code: 1
```

the 2 errors it produces

* Failure: ''
* ServiceException: 503 None

going to re-run with debug logging

**Comment by emk** [comment]
Ugh. It looks like a problem on Google's end. It's hard for us to retry one of these transfers, because `dbcrossbar` never saves the data locally, preferring to just to stream it through.

We _could_ rebuild the internals of `dbcrossbar` to allow restarting a stream after a failure occurs, but that's a major architectural change.

OTOH, if `gsutil` can't handle simple uploads without failing, we may have no choice. :-(

**Comment by emk** [comment]
Specifically, the core type of `dbcrossbar` is currently:

```rust
BoxFuture<Option<BoxStream<CsvStream>>>
```

This is:

- A **future**
- that yields an **optional**
- **stream** where each item is itself
- a **CSV data stream**.

The problem is that once we ask the `BoxStream` for a `CsvStream` and start consuming it, there's no way to restart the `CsvStream` if the destination randomly fails. To fix this, we'd need to use:

```rust
BoxFuture<Option<BoxStream<FnMut() -> Result<CsvStream>>>>
```

Here, the `BoxStream` would now return a `FnMut() -> Result<CsvStream>`, a function that _creates_ a `CsvStream` on demand.

Alternatively, we could leave that definition alone, and instead we could change the definition of `CsvStream` from:

```rust
/// A stream of CSV data, with a unique name.
pub struct CsvStream {
    /// The name of this stream.
    pub name: String,
    /// A reader associated with this stream.
    pub data: Box<dyn Stream<Item = BytesMut, Error = Error> + Send + 'static>,
}
```

...to:

```rust
/// A stream of CSV data, with a unique name.
pub struct CsvStream {
    /// The name of this stream.
    pub name: String,
    /// A reader associated with this stream.
    pub data_fn: Box<dyn FnMut() -> Result<Box<dyn Stream<Item = BytesMut, Error = Error> + Send + 'static>>>,
}
```

...replacing `data` with a `data_fn` that creates `data` on demand, making retries possible.

This represents a real increase in complexity, but it's definitely doable if we need to fix this Google issue. But is it worth the price?

**Comment by seamusabshere** [comment]
it seems to me the options (both complex) are:

1. continue to take advantage of `gsutil [-m]`, and do what you describe above;
2. or write a parallelized http downloader

The advantages of (1) are that you guarantee the fastest possible speed (I think, though see the strange bandwidth pattern we've observed when loading 500gb+ files GS->BQ). The advantage of (2) is that you can also use it for S3.

**Comment by emk** [comment]
Option (2) probably doesn't help us much, because if `gsutil cp ... -` is unreliable, I have no reason to believe that an HTTP-based client will succeed reliably. I mean, I totally want to write an HTTP downloader for `gs://` and `s3://` (using async `reqwest` and the URL-signing code we have somewhere), but that's almost certainly a separate issue.

Once we assume that either reads or writes in _any_ driver are too unreliable for typical use, then we need to build a retry mechanism that allows us to recreate a `CsvStream`. This is a couple of days of fiddly work, but totally doable.

**Comment by emk** [comment]
Actually, there is one more option here: We could write another streaming, part-wise uploader with retries for each part. This would be easier than the last time we did it for S3, because we wouldn't have to badly reinvent many features tokio provides out of the box. But that would mean ditching gsutil, etc. We want to do that, but there's a least a bit of work involved.

**Comment by seamusabshere** [comment]
i'm not sure if this is related, but here's an example of the latest version of dbcrossbar (0.2.0-alpha.8) halting after <10mb of a big data dump from gs:

```
$ dbcrossbar cp     --if-exists=overwrite   bigquery:myproject:tmp.myclient_fig_all_654534682f657c425939414c3f_20190725   csv:myclient_fig_all_6545346
82f657c425939414c3f_20190725.csv   --temporary gs://tmp   --temporary bigquery:myproject:tmp
app: dbcrossbar
 ver: 0.2.0-alpha.8
  Jul 26 15:38:56.006 DEBG Opt { log_format: Indented, log_extra: [], cmd: Cp { command: Opt { if_exists: Overwrite, schema: None, temporaries: ["gs://tmp", "bigquery:myproject:tm
p"], where_clause: None, from_locator: BigQueryLocator { table_name: TableName { project: "myproject", dataset: "tmp", table: "myclient_fig_all_654534682f657c425939414c3f_20190725" } }, to_lo
cator: CsvLocator { path: Path("myclient_fig_all_654534682f657c425939414c3f_20190725.csv") } } } }
  Jul 26 15:38:56.518 DEBG BigQuery schema: [[...]]

  Jul 26 15:38:56.519 DEBG performaning local data transfer
  from_locator: bigquery:myproject:tmp.myclient_fig_all_654534682f657c425939414c3f_20190725
   to_temp: gs://tmp/SgwV4kXAga/
    Jul 26 15:38:56.519 DEBG export SQL: SELECT [...] FROM `myproject.tmp.myclient_fig_all_654534682f657c425939414c3f_20190725`
    Jul 26 15:38:56.519 DEBG[2019-07-26T15:38:56Z DEBUG tokio_reactor] adding I/O source: 0
[2019-07-26T15:38:56Z DEBUG tokio_reactor::registration] scheduling Write for: 0
 running `bq query`
[2019-07-26T15:38:56Z DEBUG tokio_reactor] dropping I/O source: 0
[2019-07-26T15:38:56Z DEBUG tokio_reactor] adding I/O source: 4194304
[2019-07-26T15:38:56Z DEBUG tokio_reactor::registration] scheduling Read for: 0
Waiting on bqjob_r779708cc9f3ba946_0000016c2eef659d_1 ... (109s) Current status: DONE
[2019-07-26T15:40:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:40:49Z DEBUG tokio_reactor] dropping I/O source: 0
    Jul 26 15:40:49.650 DEBG deleting existing gs://tmp/SgwV4kXAga/
[2019-07-26T15:40:49Z DEBUG tokio_reactor] adding I/O source: 0
[2019-07-26T15:40:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:40:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
CommandException: 1 files/objects could not be removed.
[2019-07-26T15:40:50Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:40:50Z DEBUG tokio_reactor] dropping I/O source: 0
    Jul 26 15:40:50.232 WARN can't delete contents of gs://tmp/SgwV4kXAga/, possibly because it doesn't exist
    Jul 26 15:40:50.233 DEBG running `bq extract`
[2019-07-26T15:40:50Z DEBUG tokio_reactor] adding I/O source: 4194304
[2019-07-26T15:40:50Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:40:50Z DEBUG tokio_reactor::registration] scheduling Read for: 0
Waiting on bqjob_r3df78541ae73802f_0000016c2ef121c9_1 ... (164s) Current status: DONE
[2019-07-26T15:43:36Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:43:36Z DEBUG tokio_reactor] dropping I/O source: 0
    Jul 26 15:43:36.523 DEBG deleting export temp table: myproject:tmp.temp_myclient_fig_all_654534682f657c425939414c3f_20190725_uBppKdApJS
[2019-07-26T15:43:36Z DEBUG tokio_reactor] adding I/O source: 8388608
[2019-07-26T15:43:36Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:43:36Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2019-07-26T15:43:37Z DEBUG tokio_reactor] dropping I/O source: 0
   from_temp: gs://tmp/SgwV4kXAga/
    Jul 26 15:43:37.092 DEBG getting CSV files from gs://tmp/SgwV4kXAga/[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 12582912

[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 0
    [2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 0
Jul 26 15:43:37.092[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 16777217
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 1
 DEBG listing gs://tmp/SgwV4kXAga/**/*.csv
    stream: 000000000000
     url: gs://tmp/SgwV4kXAga/000000000000.csv
      Jul 26 15:43:37.675 DEBG streaming from gs://tmp/SgwV4kXAga/000000000000.csv using `gsutil cp`
    stream: 000000000001
     url:[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 20971522
 gs://tmp/SgwV4kXAga/000000000001.csv
      [2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 2
Jul 26 15:43:37.676 DEBG streaming from gs://tmp/SgwV4kXAga/000000000001.csv using `gsutil cp`
[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 25165827
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 3
[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 29360132
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 4
    stream: 000000000002
     url: gs://tmp/SgwV4kXAga/000000000002.csv
      Jul 26 15:43:37.678 DEBG streaming from gs://tmp/SgwV4kXAga/000000000002.csv using `gsutil cp`
      [2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 33554437
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 5
[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 37748742
    stream: 000000000003
     url: gs://tmp/SgwV4kXAga/000000000003.csv
      Jul 26 15:43:37.679 DEBG streaming from gs://tmp/SgwV4kXAga/000000000003.csv using `gsutil cp`
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 6
[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 41943047
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 7
  to_locator: csv:myclient_fig_all_654534682f657c425939414c3f_20190725.csv
   stream: 000000000000
    path: myclient_fig_all_654534682f657c425939414c3f_20190725.csv/000000000000.csv
     Jul 26 15:43:37.682 DEBG writing CSV stream to file
   stream: 000000000001
    path: myclient_fig_all_654534682f657c425939414c3f_20190725.csv/000000000001.csv
     Jul 26 15:43:37.682 DEBG writing CSV stream to file
   stream: 000000000002
    path: myclient_fig_all_654534682f657c425939414c3f_20190725.csv/000000000002.csv
     Jul 26 15:43:37.683 DEBG writing CSV stream to file
   stream: 000000000003
    path: myclient_fig_all_654534682f657c425939414c3f_20190725.csv/000000000003.csv
     Jul 26 15:43:37.683 DEBG writing CSV stream to file
[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 46137352
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 8
[2019-07-26T15:43:37Z DEBUG tokio_reactor] adding I/O source: 50331657
[2019-07-26T15:43:37Z DEBUG tokio_reactor::registration] scheduling Read for: 9
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 7
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 7
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 5
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 5
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 7
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 3
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 3
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 1
[2019-07-26T15:43:38Z DEBUG tokio_reactor] dropping I/O source: 1
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 9
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 9
[2019-07-26T15:43:38Z DEBUG tokio_reactor::registration] scheduling Read for: 9
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 8
[2019-07-26T15:43:39Z DEBUG tokio_reactor::registration] scheduling Read for: 4
```

**Comment by emk** [comment]
Let's track silent stalls and hard errors as different problems until proven otherwise.

**Comment by seamusabshere** [comment]
here's a new set of errors:

```
$ dbcrossbar cp   --if-exists=append   --schema=postgres-sql:people.sql   gs://myproject-bucket/people_final_v58/   postgres:foobar
Traceback (most recent call last):
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gsutil", line 21, in <module>
    gsutil.RunMain()
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gsutil.py", line 132, in RunMain
    sys.exit(gslib.__main__.main())
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 431, in main
    user_project=user_project)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 760, in _RunNamedCommandAndHandleExceptions
    _HandleUnknownFailure(e)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 626, in _RunNamedCommandAndHandleExceptions
    user_project=user_project)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/command_runner.py", line 411, in RunNamedCommand
    return_code = command_inst.RunCommand()
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/commands/cp.py", line 1136, in RunCommand
    cat_out_fd=cat_out_fd)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/utils/cat_helper.py", line 152, in CatUrlStrings
    provider=storage_url.scheme)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/cloud_api_delegator.py", line 353, in GetObjectMedia
    decryption_tuple=decryption_tuple)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 1145, in GetObjectMedia
    decryption_tuple=decryption_tuple)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 1270, in _PerformDownload
    use_chunks=False)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/transfer.py", line 536, in StreamMedia
    additional_headers=additional_headers)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/transfer.py", line 415, in __GetChunk
    retries=self.num_retries)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 356, in MakeRequest
    max_retry_wait, total_wait_sec))
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/utils/retry_util.py", line 84, in RetriesInDataTransferHandler
    http_wrapper.RethrowExceptionHandler(retry_args)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 346, in MakeRequest
    check_response_func=check_response_func)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 396, in _MakeRequestNoRetry
    redirections=redirections, connection_type=connection_type)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_media.py", line 546, in NewRequest
    connection_type=connection_type)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/transport.py", line 175, in new_request
    redirections, connection_type)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/transport.py", line 282, in request
    connection_type=connection_type)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/third_party/httplib2/python2/httplib2/__init__.py", line 1694, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_media.py", line 453, in OverrideRequest
    headers)
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_media.py", line 668, in _conn_request
    response = conn.getresponse()
  File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/gcs_json_media.py", line 369, in getresponse
    orig_response = http_client.HTTPConnection.getresponse(self)
  File "/usr/lib/python2.7/httplib.py", line 1121, in getresponse
    response.begin()
  File "/usr/lib/python2.7/httplib.py", line 438, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python2.7/httplib.py", line 402, in _read_status
    raise BadStatusLine(line)
httplib.BadStatusLine: ''
Error: gsutil cp gs://myproject-bucket/people_final_v58/000000001959.csv - failed with exit code: 1
```

**Comment by seamusabshere** [comment]
this appears to have been fixed along the way.
